pkg/asset/installconfig/gcp/dns.go

```diff
@@ -34,41 +40,72 @@ func GetPublicZone(ctx context.Context, project, baseDomain string) (*dns.Manage
 	return dnsZone, nil
 }
 
+// GetPublicZoneByID returns a DNS managed zone from the provided project which matches the ID.
+func GetPublicZoneByID(ctx context.Context, project, id string) (*dns.ManagedZone, error) {
+	client, err := NewClient(context.TODO())
+	if err != nil {
+		return nil, err
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Minute)
+	defer cancel()
+
+	dnsZone, err := client.GetPublicDNSZoneByID(ctx, project, id)
+	if err != nil {
+		return nil, err
+	}
+	return dnsZone, nil
+}
+
 // GetBaseDomain returns a base domain chosen from among the project's public DNS zones.
-func GetBaseDomain(project string) (string, error) {
+func GetBaseDomain(project string) (*Zone, error) {
 	client, err := NewClient(context.TODO())
 	if err != nil {
-		return "", err
+		return nil, err
 	}
 	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Minute)
 	defer cancel()
 
-	publicZones, err := client.GetPublicDomains(ctx, project)
+	publicDomains, err := client.GetPublicDomains(ctx, project)
 	if err != nil {
-		return "", errors.Wrap(err, "could not retrieve base domains")
+		return nil, errors.Wrap(err, "could not retrieve base domains")
 	}
+
+	publicZones := make([]string, 0, len(publicDomains))
+	for name := range publicDomains {
+		for _, id := range publicDomains[name] {
```

Comment:
nit: I think it would be more idiomatic/easier to read:

for name, ids := range publicDomains {
		for _, id := range ids {